File: C++/Ejercicio8.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <iomanip>
#include <algorithm>

using namespace std;

// Función auxiliar para mostrar arreglo
void mostrarArreglo(const vector<int>& arr) {
    cout << "   [";
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i];
        if (i < arr.size() - 1) cout << ", ";
    }
    cout << "]" << endl;
}

// Función auxiliar para crear líneas de separación
string crearSeparador(int longitud, char caracter = '=') {
    return string(longitud, caracter);
}

void insertarEnArreglos() {
    cout << ">>> INSERCIÓN EN ARREGLOS" << endl;
    cout << crearSeparador(75, '-') << endl;
    
    // Arreglo original
    vector<int> original = {10, 20, 30, 40};
    cout << "✓ Arreglo original: " << endl;
    mostrarArreglo(original);
    
    // Insertar al final (requiere copiar)
    cout << "\n1. INSERTAR AL FINAL:" << endl;
    vector<int> conNuevoFinal(original);
    conNuevoFinal.resize(5);
    conNuevoFinal[4] = 50;
    cout << "   Resultado: " << endl;
    mostrarArreglo(conNuevoFinal);
    cout << "   Código: resize() y asignación manual" << endl;
    cout << "   Complejidad: O(n) - requiere copiar datos" << endl;
    
    // Insertar al inicio
    cout << "\n2. INSERTAR AL INICIO:" << endl;
    vector<int> conNuevoInicio(5);
    conNuevoInicio[0] = 5;
    copy(original.begin(), original.end(), conNuevoInicio.begin() + 1);
    cout << "   Resultado: " << endl;
    mostrarArreglo(conNuevoInicio);
    cout << "   Código: Desplazamiento manual + copy()" << endl;
    cout << "   Complejidad: O(n) - desplaza todos los elementos" << endl;
    
    // Insertar en el medio
    cout << "\n3. INSERTAR EN EL MEDIO (posición 2):" << endl;
    vector<int> conNuevoMedio(5);
    copy(original.begin(), original.begin() + 2, conNuevoMedio.begin());
    conNuevoMedio[2] = 25;
    copy(original.begin() + 2, original.end(), conNuevoMedio.begin() + 3);
    cout << "   Resultado: " << endl;
    mostrarArreglo(conNuevoMedio);
    cout << "   Código: Múltiples copy() calls" << endl;
    cout << "   Complejidad: O(n) - desplaza elementos después de la posición" << endl;
    
    cout << "\n❌ Desventajas:" << endl;
    cout << "   • Tamaño fijo: hay que crear un nuevo vector/array" << endl;
    cout << "   • Código complejo: requiere múltiples operaciones" << endl;
    cout << "   • Propenso a errores: fácil equivocarse con índices" << endl;
    cout << "   • Ineficiente: copia datos constantemente" << endl;
}

void insertarEnVector() {
    cout << ">>> INSERCIÓN EN VECTOR" << endl;
    cout << crearSeparador(75, '-') << endl;
    
    // Vector original
    vector<int> original = {10, 20, 30, 40};
    cout << "✓ Vector original: ";
    for (int val : original) cout << val << " ";
    cout << endl;
    
    // Insertar al final
    cout << "\n1. INSERTAR AL FINAL:" << endl;
    vector<int> conNuevoFinal = original;
    conNuevoFinal.push_back(50);
    cout << "   Resultado: ";
    for (int val : conNuevoFinal) cout << val << " ";
    cout << endl;
    cout << "   Código: vec.push_back(50)" << endl;
    cout << "   Complejidad: O(1) amortizado - muy eficiente" << endl;
    
    // Insertar al inicio
    cout << "\n2. INSERTAR AL INICIO:" << endl;
    vector<int> conNuevoInicio = original;
    conNuevoInicio.insert(conNuevoInicio.begin(), 5);
    cout << "   Resultado: ";
    for (int val : conNuevoInicio) cout << val << " ";
    cout << endl;
    cout << "   Código: vec.insert(vec.begin(), 5)" << endl;
    cout << "   Complejidad: O(n) - desplaza elementos internamente" << endl;
    
    // Insertar en el medio
    cout << "\n3. INSERTAR EN EL MEDIO (posición 2):" << endl;
    vector<int> conNuevoMedio = original;
    conNuevoMedio.insert(conNuevoMedio.begin() + 2, 25);
    cout << "   Resultado: ";
    for (int val : conNuevoMedio) cout << val << " ";
    cout << endl;
    cout << "   Código: vec.insert(vec.begin() + 2, 25)" << endl;
    cout << "   Complejidad: O(n) - desplaza elementos después de la posición" << endl;
    
    cout << "\n✓ Ventajas:" << endl;
    cout << "   • Tamaño dinámico: redimensionamiento automático" << endl;
    cout << "   • Código simple: una línea para cualquier inserción" << endl;
    cout << "   • Seguro: vector maneja la lógica interna" << endl;
    cout << "   • Flexible: métodos para insertar en cualquier posición" << endl;
}

void analizarDesempenio() {
    cout << ">>> ANÁLISIS DE DESEMPEÑO" << endl;
    cout << crearSeparador(75, '-') << endl;
    
    int tamaños[] = {100, 1000, 10000};
    
    for (int tamaño : tamaños) {
        cout << "\nInsertando en vector con " << tamaño << " elementos:" << endl;
        
        vector<int> vec;
        for (int i = 0; i < tamaño; i++) {
            vec.push_back(i);
        }
        
        // Insertar al final
        auto inicio = chrono::high_resolution_clock::now();
        vec.push_back(9999);
        auto tiempoFinal = chrono::high_resolution_clock::now();
        auto duracionFinal = chrono::duration_cast<chrono::nanoseconds>(tiempoFinal - inicio).count();
        
        // Recrear vector
        vec.clear();
        for (int i = 0; i < tamaño; i++) {
            vec.push_back(i);
        }
        
        // Insertar al inicio
        inicio = chrono::high_resolution_clock::now();
        vec.insert(vec.begin(), -1);
        auto tiempoInicio = chrono::high_resolution_clock::now();
        auto duracionInicio = chrono::duration_cast<chrono::nanoseconds>(tiempoInicio - inicio).count();
        
        // Recrear vector
        vec.clear();
        for (int i = 0; i < tamaño; i++) {
            vec.push_back(i);
        }
        
        // Insertar al medio
        inicio = chrono::high_resolution_clock::now();
        vec.insert(vec.begin() + tamaño / 2, 5000);
        auto tiempoMedio = chrono::high_resolution_clock::now();
        auto duracionMedio = chrono::duration_cast<chrono::nanoseconds>(tiempoMedio - inicio).count();
        
        cout << "  • Al final: " << duracionFinal << " ns (RÁPIDO)" << endl;
        cout << "  • Al inicio: " << duracionInicio << " ns (LENTO)" << endl;
        cout << "  • Al medio: " << duracionMedio << " ns (MEDIO)" << endl;
    }
    
    cout << "\n📊 Conclusión:" << endl;
    cout << "  • Insertar al final: O(1) - muy rápido ✓" << endl;
    cout << "  • Insertar al inicio: O(n) - lento (desplaza todos)" << endl;
    cout << "  • Insertar al medio: O(n) - lento (desplaza mitad)" << endl;
}

void mostrarComparativaInsercion() {
    cout << ">>> TABLA COMPARATIVA: INSERCIÓN" << endl;
    cout << crearSeparador(75, '-') << endl;
    
    cout << left << setw(30) << "Operación" << " | " 
         << setw(20) << "Arreglo/Vector" << " | " 
         << setw(20) << "Vector STL" << endl;
    cout << crearSeparador(75, '-') << endl;
    
    cout << left << setw(30) << "Insertar al final" << " | " 
         << setw(20) << "Requiere resize" << " | " 
         << setw(20) << "vec.push_back()" << endl;
    cout << left << setw(30) << "Código para el final" << " | " 
         << setw(20) << "7-8 líneas" << " | " 
         << setw(20) << "1 línea" << endl;
    
    cout << left << setw(30) << "Insertar al inicio" << " | " 
         << setw(20) << "Desplazar todos" << " | " 
         << setw(20) << "vec.insert()" << endl;
    cout << left << setw(30) << "Código para el inicio" << " | " 
         << setw(20) << "10+ líneas" << " | " 
         << setw(20) << "1 línea" << endl;
    
    cout << left << setw(30) << "Insertar en el medio" << " | " 
         << setw(20) << "Desplazar parciales" << " | " 
         << setw(20) << "vec.insert()" << endl;
    cout << left << setw(30) << "Código para el medio" << " | " 
         << setw(20) << "Muy complejo" << " | " 
         << setw(20) << "1 línea" << endl;
    
    cout << left << setw(30) << "Facilidad de uso" << " | " 
         << setw(20) << "Difícil ❌" << " | " 
         << setw(20) << "Fácil ✓" << endl;
    cout << left << setw(30) << "Propenso a errores" << " | " 
         << setw(20) << "Sí ❌" << " | " 
         << setw(20) << "No ✓" << endl;
    cout << left << setw(30) << "Redimensionamiento" << " | " 
         << setw(20) << "Manual" << " | " 
         << setw(20) << "Automático" << endl;
    
    cout << "\n>>> EJEMPLO PRÁCTICO:" << endl;
    cout << crearSeparador(75, '-') << endl;
    
    cout << "\nInsertar 5 en la posición 2 de [10, 20, 30, 40]:" << endl;
    
    cout << "\n❌ CON ARREGLO MANUAL:" << endl;
    cout << "  int arr[4] = {10, 20, 30, 40};" << endl;
    cout << "  int nuevo[5];" << endl;
    cout << "  copy(arr, arr + 2, nuevo);" << endl;
    cout << "  nuevo[2] = 5;" << endl;
    cout << "  copy(arr + 2, arr + 4, nuevo + 3);" << endl;
    cout << "  // Total: 5 líneas de código complejo" << endl;
    
    cout << "\n✓ CON VECTOR:" << endl;
    cout << "  vector<int> vec = {10, 20, 30, 40};" << endl;
    cout << "  vec.insert(vec.begin() + 2, 5);" << endl;
    cout << "  // Total: 2 líneas, código claro" << endl;
    
    cout << "\n>>> RECOMENDACIÓN:" << endl;
    cout << crearSeparador(75, '-') << endl;
    cout << "Use vector para inserción frecuente en cualquier posición." << endl;
    cout << "Use arreglos solo si el tamaño es fijo y no necesita insertar." << endl;
}

int main() {
    cout << "========== COMPARACIÓN: INSERCIÓN EN ARREGLOS vs VECTOR ==========\n" << endl;
    
    insertarEnArreglos();
    
    cout << "\n" << crearSeparador(75) << "\n" << endl;
    
    insertarEnVector();
    
    cout << "\n" << crearSeparador(75) << "\n" << endl;
    
    analizarDesempenio();
    
    cout << "\n" << crearSeparador(75) << "\n" << endl;
    
    mostrarComparativaInsercion();
    
    return 0;
}

File: Java/Ejercicio8.java
import java.util.ArrayList;

public class Ejercicio8 {
    
    public static void main(String[] args) {
        System.out.println("========== COMPARACIÓN: INSERCIÓN EN ARREGLOS vs ARRAYLIST ==========\n");
        
        // Inserción en Arreglos
        insertarEnArreglos();
        
        System.out.println("\n" + "=".repeat(75) + "\n");
        
        // Inserción en ArrayList
        insertarEnArrayList();
        
        System.out.println("\n" + "=".repeat(75) + "\n");
        
        // Análisis de desempeño
        analizarDesempenio();
        
        System.out.println("\n" + "=".repeat(75) + "\n");
        
        // Comparativa detallada
        mostrarComparativaInsercion();
    }
    
    // Demostración de inserción en Arreglos
    static void insertarEnArreglos() {
        System.out.println(">>> INSERCIÓN EN ARREGLOS");
        System.out.println("-".repeat(75));
        
        // Arreglo original
        int[] original = {10, 20, 30, 40};
        System.out.println("✓ Arreglo original: ");
        mostrarArreglo(original);
        
        // Insertar al final (fácil)
        System.out.println("\n1. INSERTAR AL FINAL:");
        int[] conNuevoFinal = new int[5];
        System.arraycopy(original, 0, conNuevoFinal, 0, original.length);
        conNuevoFinal[4] = 50;
        System.out.println("   Resultado: ");
        mostrarArreglo(conNuevoFinal);
        System.out.println("   Código: System.arraycopy() y asignación manual");
        System.out.println("   Complejidad: O(n) - requiere copiar todo el arreglo");
        
        // Insertar al inicio (difícil)
        System.out.println("\n2. INSERTAR AL INICIO:");
        int[] conNuevoInicio = new int[5];
        conNuevoInicio[0] = 5;
        System.arraycopy(original, 0, conNuevoInicio, 1, original.length);
        System.out.println("   Resultado: ");
        mostrarArreglo(conNuevoInicio);
        System.out.println("   Código: Desplazamiento manual + System.arraycopy()");
        System.out.println("   Complejidad: O(n) - desplaza todos los elementos");
        
        // Insertar en el medio (muy difícil)
        System.out.println("\n3. INSERTAR EN EL MEDIO (posición 2):");
        int[] conNuevoMedio = new int[5];
        System.arraycopy(original, 0, conNuevoMedio, 0, 2);
        conNuevoMedio[2] = 25;
        System.arraycopy(original, 2, conNuevoMedio, 3, 2);
        System.out.println("   Resultado: ");
        mostrarArreglo(conNuevoMedio);
        System.out.println("   Código: Múltiples System.arraycopy() calls");
        System.out.println("   Complejidad: O(n) - desplaza elementos después de la posición");
        
        System.out.println("\n❌ Desventajas:");
        System.out.println("   • Tamaño fijo: hay que crear un nuevo arreglo");
        System.out.println("   • Código complejo: requiere múltiples operaciones");
        System.out.println("   • Propenso a errores: fácil equivocarse con índices");
        System.out.println("   • Ineficiente: copia datos constantemente");
    }
    
    // Demostración de inserción en ArrayList
    static void insertarEnArrayList() {
        System.out.println(">>> INSERCIÓN EN ARRAYLIST");
        System.out.println("-".repeat(75));
        
        // ArrayList original
        ArrayList<Integer> original = new ArrayList<>();
        original.add(10);
        original.add(20);
        original.add(30);
        original.add(40);
        System.out.println("✓ ArrayList original: " + original);
        
        // Insertar al final (fácil)
        System.out.println("\n1. INSERTAR AL FINAL:");
        ArrayList<Integer> conNuevoFinal = new ArrayList<>(original);
        conNuevoFinal.add(50);
        System.out.println("   Resultado: " + conNuevoFinal);
        System.out.println("   Código: list.add(50)");
        System.out.println("   Complejidad: O(1) amortizado - muy eficiente");
        
        // Insertar al inicio (fácil)
        System.out.println("\n2. INSERTAR AL INICIO:");
        ArrayList<Integer> conNuevoInicio = new ArrayList<>(original);
        conNuevoInicio.add(0, 5);
        System.out.println("   Resultado: " + conNuevoInicio);
        System.out.println("   Código: list.add(0, 5)");
        System.out.println("   Complejidad: O(n) - desplaza elementos internamente");
        
        // Insertar en el medio (fácil)
        System.out.println("\n3. INSERTAR EN EL MEDIO (posición 2):");
        ArrayList<Integer> conNuevoMedio = new ArrayList<>(original);
        conNuevoMedio.add(2, 25);
        System.out.println("   Resultado: " + conNuevoMedio);
        System.out.println("   Código: list.add(2, 25)");
        System.out.println("   Complejidad: O(n) - desplaza elementos después de la posición");
        
        System.out.println("\n✓ Ventajas:");
        System.out.println("   • Tamaño dinámico: redimensionamiento automático");
        System.out.println("   • Código simple: una línea para cualquier inserción");
        System.out.println("   • Seguro: ArrayList maneja la lógica interna");
        System.out.println("   • Flexible: métodos para insertar en cualquier posición");
    }
    
    // Análisis de desempeño
    static void analizarDesempenio() {
        System.out.println(">>> ANÁLISIS DE DESEMPEÑO");
        System.out.println("-".repeat(75));
        
        int[] tamaños = {100, 1000, 10000};
        
        for (int tamaño : tamaños) {
            System.out.println("\nInsertando en ArrayList con " + tamaño + " elementos:");
            
            ArrayList<Integer> list = new ArrayList<>();
            for (int i = 0; i < tamaño; i++) {
                list.add(i);
            }
            
            // Insertar al final
            long inicio = System.nanoTime();
            list.add(9999);
            long tiempoFinal = System.nanoTime() - inicio;
            
            // Insertar al inicio
            inicio = System.nanoTime();
            list.add(0, -1);
            long tiempoInicio = System.nanoTime() - inicio;
            
            // Insertar al medio
            inicio = System.nanoTime();
            list.add(tamaño / 2, 5000);
            long tiempoMedio = System.nanoTime() - inicio;
            
            System.out.printf("  • Al final: %,d ns (RÁPIDO)\n", tiempoFinal);
            System.out.printf("  • Al inicio: %,d ns (LENTO)\n", tiempoInicio);
            System.out.printf("  • Al medio: %,d ns (MEDIO)\n", tiempoMedio);
        }
        
        System.out.println("\n📊 Conclusión:");
        System.out.println("  • Insertar al final: O(1) - muy rápido ✓");
        System.out.println("  • Insertar al inicio: O(n) - lento (desplaza todos)");
        System.out.println("  • Insertar al medio: O(n) - lento (desplaza mitad)");
    }
    
    // Comparativa detallada
    static void mostrarComparativaInsercion() {
        System.out.println(">>> TABLA COMPARATIVA: INSERCIÓN");
        System.out.println("-".repeat(75));
        
        System.out.println(String.format("%-30s | %-20s | %-20s", "Operación", "Arreglo", "ArrayList"));
        System.out.println("-".repeat(75));
        
        System.out.println(String.format("%-30s | %-20s | %-20s", "Insertar al final", "Requiere nuevo array", "list.add(value)"));
        System.out.println(String.format("%-30s | %-20s | %-20s", "Código para el final", "7-8 líneas", "1 línea"));
        
        System.out.println(String.format("%-30s | %-20s | %-20s", "Insertar al inicio", "Desplazar todos", "list.add(0, value)"));
        System.out.println(String.format("%-30s | %-20s | %-20s", "Código para el inicio", "10+ líneas", "1 línea"));
        
        System.out.println(String.format("%-30s | %-20s | %-20s", "Insertar en el medio", "Desplazar parciales", "list.add(idx, value)"));
        System.out.println(String.format("%-30s | %-20s | %-20s", "Código para el medio", "Muy complejo", "1 línea"));
        
        System.out.println(String.format("%-30s | %-20s | %-20s", "Facilidad de uso", "Difícil ❌", "Fácil ✓"));
        System.out.println(String.format("%-30s | %-20s | %-20s", "Propenso a errores", "Sí ❌", "No ✓"));
        System.out.println(String.format("%-30s | %-20s | %-20s", "Redimensionamiento", "Manual", "Automático"));
        System.out.println(String.format("%-30s | %-20s | %-20s", "Complejidad de tiempo", "O(n)", "O(n)"));
        System.out.println(String.format("%-30s | %-20s | %-20s", "Complejidad espacial", "O(n)", "O(n)"));
        
        System.out.println("\n>>> EJEMPLO PRÁCTICO:");
        System.out.println("-".repeat(75));
        
        System.out.println("\nInsertar 5 en la posición 2 de [10, 20, 30, 40]:");
        
        System.out.println("\n❌ CON ARREGLO:");
        System.out.println("  int[] original = {10, 20, 30, 40};");
        System.out.println("  int[] nuevo = new int[5];");
        System.out.println("  System.arraycopy(original, 0, nuevo, 0, 2);");
        System.out.println("  nuevo[2] = 5;");
        System.out.println("  System.arraycopy(original, 2, nuevo, 3, 2);");
        System.out.println("  // Total: 5 líneas de código complejo");
        
        System.out.println("\n✓ CON ARRAYLIST:");
        System.out.println("  ArrayList<Integer> list = new ArrayList<>();");
        System.out.println("  list.addAll(Arrays.asList(10, 20, 30, 40));");
        System.out.println("  list.add(2, 5);");
        System.out.println("  // Total: 3 líneas, código claro");
        
        System.out.println("\n>>> RECOMENDACIÓN:");
        System.out.println("-".repeat(75));
        System.out.println("Use ArrayList para inserción frecuente en cualquier posición.");
        System.out.println("Use arreglos solo si el tamaño es fijo y no necesita insertar.");
    }
    
    // Método auxiliar para mostrar arreglo
    static void mostrarArreglo(int[] arr) {
        System.out.print("   [");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            if (i < arr.length - 1) System.out.print(", ");
        }
        System.out.println("]");
    }
}
